### PR TITLE
Add ModularProduct

### PR DESCRIPTION
**Pull request by james-d-mitchell**
This PR partially introduces the `ModularProduct` of two digraphs, as used in @le27's PR #334. When trying to understand the code in PR #334, I wrote the method in this PR. It seems to be about 10 times faster than the method in PR #334 for finding the modular product, but it doesn't do exactly the same thing. 

 @le27 can I suggest that you finish this PR off (add tests and documentation), and then use the `ModularProduct` in your other PR #334? 

It's possible that it would be beneficial to have a `ReducedModularProduct` operation too (which basically returns the same thing as `ReducedDigraph(ModularProduct(D));` but without the creation of the intermediate (possibly too large) `ModularProduct(D)`). Having said that I don't think the time consuming part of `ReducedDigraph(ModularProduct(D));` is the call to `ReducedDigraph` so this might also not be worth it.

### CODE REVIEW

File: gap/oper.gi [hunk 1]
```diff
@@ -612,6 +612,51 @@ function(arg)
...
+      for w in OutNeighbours(E1)[u] do
+        for x in OutNeighbours(E2)[v] do
+          if (u = w) = (v = x) then
+            Add(edges[next], map(w, x));
+          fi;
+        od;
+      od;
+    od;
+  od;
+  return Digraph(edges);
```
Comment (james-d-mitchell) [inline gap/oper.gi:657]:
```suggestion
  return DigraphNC(edges);
```

File: gap/oper.gi [hunk 2]
```diff
@@ -612,6 +612,51 @@ function(arg)
   return D;
 end);
 
+InstallMethod(ModularProduct, "for a digraph and digraph",
+[IsDigraph, IsDigraph],
+function(D1, D2)
+  local m, n, E1, E2, edges, next, map, u, v, w, x;
+
+  if IsMultiDigraph(D1) or IsMultiDigraph(D2) then
+    ErrorNoReturn("ModularProduct does not support multidigraphs.");
```
Comment (james-d-mitchell) [inline gap/oper.gi:621]:
```suggestion
    ErrorNoReturn("ModularProduct does not support multidigraphs,");
```

File: doc/oper.xml [hunk 3]
```diff
@@ -1740,3 +1740,42 @@ gap> DigraphDijkstra(D, 1, 2);
...
+  <Description>
+  If <A>D1</A> and <A>D2</A> are digraphs without multiedges, then
+  <C>ModularProduct</C> calculates the Modular Product Digraph <C>MPD</C> of
+  <A>D1</A>and <A>D2</A>.
+
+  <C>MPD</C> has vertex set <C>V1 x V2</C> where <C>V1</C> is the vertex set of
+  <A>D1</A> and <C>V2</C> is the vertex set of <A>D2</A> (a vertex
+  <C>[a, b]</C> has label <C>(a - 1) * |V2| + b</C> in the  output). There is
+  an edge from <C>[a, b]</C> to <C>[c, d]</C> precisely if both of:
+
```
Comment (james-d-mitchell) [inline doc/oper.xml:1758]:
```suggestion
<List>
```

File: doc/oper.xml [hunk 4]
```diff
@@ -1740,3 +1740,42 @@ gap> DigraphDijkstra(D, 1, 2);
...
+  1. The vertices <C>a</C> and <C>c</C> of <A>D1</A> are equal if and only if
```
Comment (james-d-mitchell) [inline doc/oper.xml:1759]:
```suggestion
  <Item> The vertices <C>a</C> and <C>c</C> of <A>D1</A> are equal if and only if
```

File: doc/oper.xml [hunk 5]
```diff
@@ -1740,3 +1740,42 @@ gap> DigraphDijkstra(D, 1, 2);
...
+     the vertices <C>b</C> and <C>d</C> of <A>D2</A> are equal.
```
Comment (james-d-mitchell) [inline doc/oper.xml:1760]:
```suggestion
     the vertices <C>b</C> and <C>d</C> of <A>D2</A> are equal.</Item>
```

File: doc/oper.xml [hunk 6]
```diff
@@ -1740,3 +1740,42 @@ gap> DigraphDijkstra(D, 1, 2);
...
+
+  2. There is an edge from <C>a</C> to <C>c</C> in <A>D1</A> if and only if
```
Comment (james-d-mitchell) [inline doc/oper.xml:1762]:
```suggestion
  <Item> There is an edge from <C>a</C> to <C>c</C> in <A>D1</A> if and only if
```

File: doc/oper.xml [hunk 7]
```diff
@@ -1740,3 +1740,42 @@ gap> DigraphDijkstra(D, 1, 2);
...
+     there is an edge from <C>b</C> to <C>d</C> in <A>D2</A>.
```
Comment (james-d-mitchell) [inline doc/oper.xml:1763]:
```suggestion
     there is an edge from <C>b</C> to <C>d</C> in <A>D2</A>.</Item></List>
```

**Comment by james-d-mitchell** [comment]
I think this is good to go.

File: doc/oper.xml [hunk 8]
```diff
@@ -1740,3 +1740,42 @@ gap> DigraphDijkstra(D, 1, 2);
...
+  <Returns>A digraph.</Returns>
+  <Description>
+  If <A>D1</A> and <A>D2</A> are digraphs without multiedges, then
+  <C>ModularProduct</C> calculates the Modular Product Digraph <C>MPD</C> of
+  <A>D1</A>and <A>D2</A>.
+
+  <C>MPD</C> has vertex set <C>V1 x V2</C> where <C>V1</C> is the vertex set of
+  <A>D1</A> and <C>V2</C> is the vertex set of <A>D2</A> (a vertex
+  <C>[a, b]</C> has label <C>(a - 1) * |V2| + b</C> in the  output). There is
+  an edge from <C>[a, b]</C> to <C>[c, d]</C> precisely if both of:
```
Comment (wilfwilson) [inline doc/oper.xml:1757]:
```suggestion
  an edge from <C>[a, b]</C> to <C>[c, d]</C> precisely when the following
  two conditions are satisfied:
```

File: doc/oper.xml [hunk 9]
```diff
@@ -1740,3 +1740,42 @@ gap> DigraphDijkstra(D, 1, 2);
...
+<List>
+  <Item> The vertices <C>a</C> and <C>c</C> of <A>D1</A> are equal if and only if
+     the vertices <C>b</C> and <C>d</C> of <A>D2</A> are equal.</Item>
+
+  <Item> There is an edge from <C>a</C> to <C>c</C> in <A>D1</A> if and only if
+     there is an edge from <C>b</C> to <C>d</C> in <A>D2</A>.</Item></List>
```
Comment (wilfwilson) [inline doc/oper.xml:1763]:
```suggestion
     there is an edge from <C>b</C> to <C>d</C> in <A>D2</A>.</Item>
   </List>
```

File: doc/oper.xml [hunk 10]
```diff
@@ -1740,3 +1740,42 @@ gap> DigraphDijkstra(D, 1, 2);
...
+
+  are satisfied. This digraph is of note as the complete (with loops) subgraphs
+  of <C>MPD</C> are precisely the partial isomorphisms from <A>D1</A> to
+  <A>D2</A>.
```
Comment (wilfwilson) [inline doc/oper.xml:1767]:
```suggestion
  Notably, the complete (with loops) subdigraphs of <C>MPD</C> are
  precisely the partial isomorphisms from <A>D1</A> to <A>D2</A>.
```

File: doc/oper.xml [hunk 11]
```diff
@@ -1740,3 +1740,42 @@ gap> DigraphDijkstra(D, 1, 2);
...
+  <C>[a, b]</C> has label <C>(a - 1) * |V2| + b</C> in the  output). There is
+  an edge from <C>[a, b]</C> to <C>[c, d]</C> precisely when the following
+  two conditions are satisfied:
+<List>
+  <Item> The vertices <C>a</C> and <C>c</C> of <A>D1</A> are equal if and only if
+     the vertices <C>b</C> and <C>d</C> of <A>D2</A> are equal.</Item>
+
+  <Item> There is an edge from <C>a</C> to <C>c</C> in <A>D1</A> if and only if
+     there is an edge from <C>b</C> to <C>d</C> in <A>D2</A>.</Item>
+   </List>
```
Comment (wilfwilson) [inline doc/oper.xml:1764]:
```suggestion
  <List>
    <Item>
      The vertices <C>a</C> and <C>c</C> of <A>D1</A> are equal if and only if
      the vertices <C>b</C> and <C>d</C> of <A>D2</A> are equal.
    </Item>
    <Item>
      There is an edge from <C>a</C> to <C>c</C> in <A>D1</A> if and only if
      there is an edge from <C>b</C> to <C>d</C> in <A>D2</A>.
    </Item>
  </List>
```

**Review by wilfwilson — APPROVED** [review]

**Comment by james-d-mitchell** [comment]
Thanks @wilfwilson !

**Pull request merged**